### PR TITLE
Make source code for notebook containers public

### PR DESCRIPTION
**Pull request by mbaijal**
**Which issue is resolved by this Pull Request:**
Make the source code for the Jupyter Images to be used in Kubeflow Notebooks public. 

**Description of your changes:**
Adds - 
 - [x] Dockerfile and requirements.txt for Tensorflow CPU and GPU
 - [x] Dockerfile and requirements.txt for Pytorch CPU and GPU
 - [x] A Readme file. 
 
Tests - https://gist.github.com/mbaijal/a856a3b6cbb77b08975f6cd4e6e8b138

**TODO:**
 - [x] requirements.txt is similar - can create a common source
 - [ ] automate build (separate PR)
 - [ ] add a test (separate PR)

**Testing:**
- [x] Unit tests pass
- [x] e2e tests pass
- Details about new tests (If this PR adds a new feature)
- Details about any manual tests performed

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: components/notebook-dockerfiles/pytorch/cpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  wget \
+  zip \
+  emacs \
+  python3-pip \
+  python3-dev \
+  python3-setuptools \
+  && apt-get clean && \
+  rm -rf /var/lib/apt/lists/*
+
+# install -- node.js
```
Comment (mbaijal) [inline components/notebook-dockerfiles/pytorch/cpu/Dockerfile:49]: @mbaijal Check if this is still needed

File: components/notebook-dockerfiles/pytorch/cpu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ && chown -R ${NB_USER}:users /usr/local/bin
+
+# Install Tini - used as entrypoint for container
+RUN cd /tmp && \
+    wget --quiet https://github.com/krallin/tini/releases/download/v0.18.0/tini && \
+    echo "12d20136605531b09a2c2dac02ccee85e1b874eb322ef6baf7561cd93f93c855 *tini" | sha256sum -c - && \
+    mv tini /usr/local/bin/tini && \
+    chmod +x /usr/local/bin/tini
+
+# Install Docker Client
```
Comment (mbaijal) [inline components/notebook-dockerfiles/pytorch/cpu/Dockerfile:87]: @mbaijal This can be removed.

File: components/notebook-dockerfiles/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Using Kubeflow Notebooks on AWS
```
Comment (goswamig) [inline components/notebook-dockerfiles/README.md:1]: This title doesnt comprehend well. 

I believe you're stating using `aws optimized container images for kubeflow notebook` or along this line.

File: components/notebook-dockerfiles/pytorch/cpu/Dockerfile (hunk 2 above)
Comment (goswamig) [reply components/notebook-dockerfiles/pytorch/cpu/Dockerfile:87]: why its still here then ?

Comment (goswamig) [reply components/notebook-dockerfiles/pytorch/cpu/Dockerfile:87]: can u please fix it?

File: components/notebook-dockerfiles/pytorch/cpu/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+# Extensions for jupyter notebook
+widgetsnbextension==3.5.1
+ipywidgets==7.6.5
+
+# Kubeflow Related Packages
+kfp==1.8.6
+kfp-server-api==1.7.1
+kfserving==0.6.1
+ml-metadata==1.7.0
```
Comment (goswamig) [inline components/notebook-dockerfiles/pytorch/cpu/requirements.txt:18]: why are adding this package here?

File: components/notebook-dockerfiles/pytorch/cpu/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+matplotlib==3.4.3
+xgboost==1.5.0
+ipympl==0.8.2
+
+# AWS related packages
+awscli==1.21.5
+boto3==1.19.5
+
+# pytorch packages
+#torchelastic==0.2.2 this currently causes a dependency conflict, should be fixed very soon
```
Comment (goswamig) [inline components/notebook-dockerfiles/pytorch/cpu/requirements.txt:32]: Torchelastic is not a separate package any more, you can remove this line.

**Review by goswamig — CHANGES_REQUESTED** [review]
If requirements files are same can we write it once? Hard to maintain four copies of it.

**Comment by mbaijal** [comment]
> If requirements files are same can we write it once? Hard to maintain four copies of it.

I combined the cpu and gpu ones for pytorch and tensorflow but it is best to keep two separate copies since 
1. they have a couple of different packages and this might further change in the future. 
2. Also, in order to combine them into one, I will have to put all four Dockerfiles in the same directory (COPY src needs to be in the docker build context), which is not clean.

**Review by goswamig — APPROVED** [review]

File: components/notebook-dockerfiles/pytorch/cpu.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+# Create NB_USER user with UID=1000 and in the 'users' group
+# but allow for non-initial launches of the notebook to have
+# $HOME provided by the contents of a PV
+RUN useradd -M -s /bin/bash -N -u ${NB_UID} ${NB_USER} \
+ && mkdir -p ${HOME} \
+ && chown -R ${NB_USER}:users ${HOME} \
+ && chown -R ${NB_USER}:users /usr/local/bin
+
+# Install Tini - used as entrypoint for container
+RUN cd /tmp && \
```
Comment (mbaijal) [inline components/notebook-dockerfiles/pytorch/cpu.Dockerfile:81]: @mbaijal  Check if tini is preinstalled and this step can be skipped

File: components/notebook-dockerfiles/tensorflow/cpu.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  openssh-client \
+  sudo \
+  unzip \
+  vim \
+  wget \
+  zip \
+  emacs \
+  python3-pip \
+  python3-dev \
+  python3-setuptools \
```
Comment (akartsky) [inline components/notebook-dockerfiles/tensorflow/cpu.Dockerfile:42]: [nit] Can we put all of these in alphabetically sorted order 

(same comment for other dockerfiles too)

File: components/notebook-dockerfiles/tensorflow/cpu.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ && echo "deb https://deb.nodesource.com/node_14.x focal main" > /etc/apt/sources.list.d/nodesource.list \
+ && apt-get -yq update \
+ && apt-get -yq install --no-install-recommends \
+    nodejs \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+# Install kubectl client
+RUN echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" | sudo tee -a /etc/apt/sources.list.d/kubernetes.list && \
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add - && \
```
Comment (akartsky) [inline components/notebook-dockerfiles/tensorflow/cpu.Dockerfile:58]: [nit] Can we put the && in the next line 

(same comment for other dockerfiles too)

File: components/notebook-dockerfiles/tensorflow/cpu.Dockerfile [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+# but allow for non-initial launches of the notebook to have
+# $HOME provided by the contents of a PV
+RUN useradd -M -s /bin/bash -N -u ${NB_UID} ${NB_USER} \
+ && mkdir -p ${HOME} \
+ && chown -R ${NB_USER}:users ${HOME} \
+ && chown -R ${NB_USER}:users /usr/local/bin
+
+# Install Tini - used as entrypoint for container
+RUN cd /tmp && \
+    wget --quiet https://github.com/krallin/tini/releases/download/v0.18.0/tini && \
```
Comment (akartsky) [inline components/notebook-dockerfiles/tensorflow/cpu.Dockerfile:79]: [nit] Can we put the && in the next line

(same comment for other dockerfiles too)

**Review by akartsky — APPROVED** [review]

**Pull request merged**